### DISCUSSION TITLE
How to use the deep parameter

### DISCUSSION THREAD

**Discussion by st11x** [directus/directus#13459] [Archive]
Hi,

The docs gave one example of how to use the deep parameter to filter translations. I tried using it but instead of giving me exactly the parent that satisfies the filter, it gave me everything as shown in this screenshot.

![image](https://user-images.githubusercontent.com/7398499/169600758-34f22913-b502-4753-871d-9a95f95a0210.png)

How do I get just that one item?

Also, what other parameters are there besides "_filter" ? 

thanks
  Matt

**Comment by st11x** [comment]
I read the documentation carefully and it looks like you need to add a filter on the parent on top of adding the deep query. The deep query only affects the relations, not itself. 

So instead of this

```
?fields[]=translations.title
&deep[translations][_filter][languages_code][_eq]=fr-FR
```
do this

```
?fields[]=translations.title
&filter[translations][languages_code][_eq]=fr-FR    
&deep[translations][_filter][languages_code][_eq]=fr-FR
```

But just as I thought I had figured it out, it doesn't work again when things are slightly more nested. Is there a limit to "deep"? 

![image](https://user-images.githubusercontent.com/7398499/169669354-7494473c-e042-4cb4-acd0-5240c6299d4c.png)

**Reply by st11x** [reply]
I got a hint from this discussion [6955](https://github.com/directus/directus/discussions/6955) and was able to filter what I needed. It gets messy very quicky with M2A and a few levels deep. I tried turning up the log_level but it didn't quite help to understand what is going on. In many cases, if you made a mistake in the deep parameters, then it'll just not do that filter query and not output anything to the logs.

With just 2 different item types, I already had to do something like that

`    
        &deep[block][item:hero][translations][_filter][languages_code][_eq]=en-US
`
`
        &deep[block][item:paragraph][translations][_filter][languages_code][_eq]=en-US
`

I still have a couple more types and one more level to filter.

I had assumed that the "deep" parameter is recursive and it would likely have a limit on how deep it would go down the path, but deep is only just the next level down from the current item.

Hopefully [6654](https://github.com/directus/directus/discussions/6654) gets incorporated into the product. It looks like it will make this a lot easier to work with.

**Comment by rijkvanzanten** [comment]
> How do I get just that one item?

The usage of `deep` vs `filter` depends on what you're trying to do. If you want to fetch all top level items where a nested condition is met, use `filter` with nested fields, for example:

```
# Get all articles that have a French translation
GET /items/articles
  ?filter[translations][language_code][_eq]=fr-FR
```

however, if you'd like to get all articles regardless, but only retrieve the French translations for each article, use `deep`:

```
GET /items/articles
  ?deep[translations][_filter][language_code][_eq]=fr-FR
```

You're correct in stating that `deep` doesn't automatically apply the query recursively

> Also, what other parameters are there besides "_filter" ?

Every other "regular" query parameter, including `_limit`, `_sort`, etc 👍🏻 

---

You're correct in your discovery that `deep` doesn't automatically applies everything recursively. This is because the `deep` applies a query on the sub-collection directly, which means that it's not immediately applicable to every sub-collection that's pulled in. If you're nesting both `translations` and `categories` for example, you'd fail the query on `categories` as that wouldn't have a `language_code` field 👍🏻
